### DISCUSSION TITLE
Streaming Raw RTP data

### DISCUSSION THREAD

**Discussion by PetrFrn** [paullouisageneau/libdatachannel#517] [Q&A]
Hello,

i was looking for a solution how to stream RTP data received on a UDP port and stream it to browser app via webRTC. As a startpoint I used the example media to capture webcam media and stream it to localhost:5000. On the other site I used sfu-example and modified the receiver part and start a udp server listening on port 5000, The received data are transformed to rtc::binary and I create a rtc::message from it. Afterwards I send a message to rtc::track. But there is no video in the browser.

Probably I missed some information or I am creating incorrectly the rtc::message.

```

auto pc = std::make_shared<Receiver>();

..
pc->track = pc->conn->addTrack(media);
pc->conn->setLocalDescription();

..

n = recvfrom(sockfd, (char *)buffer, MAXLINE, MSG_WAITALL,
				             (struct sockaddr *)&cliaddr, &len);

std::byte *data = (std::byte *)buffer;
auto binaries = rtc::binary();
for (auto idx = 0; idx < n; idx++) {
	binaries.push_back(data[idx]);
};
auto message = rtc::make_message(binaries); 


pc->track->send(*message);
```

The second step would be to substitute the "sending" site with external system, that will provide me a RTP stream (audio/video) on defined ports with a known aduio and videio settings.

Thank you for you opinion.
Petr

**Comment by paullouisageneau** [comment]
Actually, you shouldn't create an `rtc::message`, which is actually an internal utility exposed only for historical reasons (it still needs to be refactored out of the streamer example). It's actually useless overhead here, as `Track` will accept the `binary` object as-is.

Also, note the flag `MSG_WAITALL` has no effect on datagram sockets, and you should check the return value of `recvfrom`. I don't know what `MAXLINE` is, but if it is set for a line of text, it might be too short a the UDP packet might be clamped as a result.

```
const char bufferSize = 2048;
char buffer[bufferSize];
n = recvfrom(sockfd, buffer, bufferSize, 0, (struct sockaddr *)&cliaddr, &len);
if(n < 0) {
     // handle error
}

pc->track->send(rtc::binary(reinterpret_cast<const rtc::byte *>(buffer), n));
```
